BootloaderCommonPkg/Library/LiteFvLib/FvLib.c

```diff
@@ -48,31 +105,97 @@ GetFfsFileByName (
+    //
+    Offset   = (UINTN)CurrentFile - (UINTN)FvHeader + GET_FFS_LENGTH (CurrentFile);
+    NextFile = (EFI_FFS_FILE_HEADER *)((UINTN)FvHeader + ALIGN_UP (Offset, 8));
+    if ((NextFile <= CurrentFile) || (NextFile > EndFile)) {
+      return EFI_ABORTED;
+    } else {
+      CurrentFile = NextFile;
+    }
   }
 
   //
-  // Get the first file
+  // File not found in this FV.
   //
-  CurrentFile = (EFI_FFS_FILE_HEADER *) ((UINTN)FvHeader + FvHeader->HeaderLength);
+  *File = NULL;
+  return EFI_NOT_FOUND;
+}
+
+
+/**
+  Find a file within a volume by its type.
+
+  This service searches for files with a specific name, within
+  either the specified firmware volume or all firmware volumes.
+
+  @param[in]       FvHeader    Pointer to a firmware volume header
+  @param[in]       FileType    The file type to search for.
+  @param[in]       Instance    0 based instance of the matched file.
+  @param[in, out]  File        Return pointer.  In the case of an error,
+                               contents are undefined.
+
+  @retval EFI_SUCCESS             The function completed successfully.
+  @retval EFI_ABORTED             An error was encountered.
+  @retval EFI_NOT_FOUND           File not found.
+  @retval EFI_INVALID_PARAMETER   One of the parameters was NULL.
+
+**/
+EFI_STATUS
+EFIAPI
+GetFfsFileByType (
+  IN  EFI_FIRMWARE_VOLUME_HEADER      *FvHeader,
+  IN  EFI_FV_FILETYPE                  FileType,
+  IN  UINT32                           Instance,
+  IN  OUT EFI_FFS_FILE_HEADER        **File
+  )
+{
+  EFI_FFS_FILE_HEADER             *CurrentFile;
+  EFI_FFS_FILE_HEADER             *NextFile;
+  EFI_FFS_FILE_HEADER             *EndFile;
+  UINT32                           Count;
+  UINT32                           Offset;
+
+  //
+  // Verify library has been initialized.
+  //
+  if (FvHeader == NULL || File == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
 
   //
   // Loop as long as we have a valid file
   //
-  EndFile = (EFI_FFS_FILE_HEADER *) (UINTN) ((UINTN)FvHeader + FvHeader->FvLength) - 1;
+  Count = 0;
+  EndFile     = (EFI_FFS_FILE_HEADER *) (UINTN) ((UINTN)FvHeader + FvHeader->FvLength) - 1;
+  CurrentFile = GetFirstFfsFileInFv (FvHeader);
   while (CurrentFile != NULL) {
-    if (CompareGuid (&CurrentFile->Name, FileName)) {
-      *File = CurrentFile;
-      return EFI_SUCCESS;
+    if (CurrentFile->Type == FileType) {
+      if (Count == Instance) {
+        *File = CurrentFile;
+        return EFI_SUCCESS;
+      }
+      Count++;
     }
 
     //
     // Get next file, compensate for 8 byte alignment if necessary.
     //
-    NextFile = (EFI_FFS_FILE_HEADER *) ((((UINTN)CurrentFile - (UINTN)FvHeader + \
-                                          GET_FFS_LENGTH (CurrentFile) + 0x07) & (~ (UINTN)7)) + (UINTN)FvHeader);
-
+    Offset   = (UINTN)CurrentFile - (UINTN)FvHeader + GET_FFS_LENGTH (CurrentFile);
+    NextFile = (EFI_FFS_FILE_HEADER *)((UINTN)FvHeader + ALIGN_UP (Offset, 8));
```

Comment:
How about using "NextFile = (EFI_FFS_FILE_HEADER *) ALIGN_UP ((UINTN)CurrentFile + GET_FFS_LENGTH (CurrentFile), 8));" instead of 2 lines above?

Reply:
The logic is not the same.   The original logic does not assume the FV header is 16 byte aligned in the location.  It means the function will work even though the FV is misaligned.

Reply:
Make sense in case FvHeader is not in aligned address.